docs/design/sum_types.md

```diff
@@ -0,0 +1,231 @@
+
+```carbon
+choice Optional(T:! Type) {
+  Some(value: T),
+  None
+}
+```
+
+A value of a function-like alternative is specified by "calling" it like a
+function, and a value of an empty alternative like `None` is specified by naming
+it:
+
+```carbon
+var my_opt: Optional(i32) = Optional(i32).None;
+my_opt = Optional(i32).Some(42);
+```
+
+The value of a choice type can be inspected using a `match` statement:
+
+```carbon
+match (my_opt) {
+  case .Some(the_value: i32) => {
+    Print(the_value);
+  }
+  case .None => {
+    Print("None");
+  }
+}
+```
+
+## User-defined sum types
+
+`choice` declarations are a convenience shorthand for common use cases, but they
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using an
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants, and the internal storage
+layout will presumably involve untagged unions or some other low-level storage
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+It does so by implementing the `Match` interface, which is defined as follows:
+
+```carbon
+interface Match {
+  let template Continuation:! Type;
+  fn Op[me: Self, C:! Continuation](continuation: Continuation*)
+    -> Continuation.ReturnType;
+}
+```
+
+`Continuation` must itself be an interface, and its definition specifies the set
+of possible alternatives: each alternative is represented as a method of that
+interface. When compiling a proper pattern (or set of patterns that includes a
+proper pattern, as with the cases of a `match`) whose type is a sum type, the
+compiler generates an implementation of `Continuation` and passes it to
+`Match.Op`. The sum type's implementation of `Match.Op` is responsible for
+determining which alternative is present and what its parameters are, and
+calling the corresponding method of `continuation` with those parameters. It is
+required to call exactly one such method exactly once before returning. The
```

Comment:
We're jumping back and forth between the compiler's responsibilities and the `Match.Op` impl's responsibilities here, so I think it'd be useful to be explicit, even though this is the most recent possible referent:
```suggestion
calling the corresponding method of `continuation` with those parameters. The `Match.Op` implementation is
required to call exactly one such method exactly once before returning. The
```